File: go/sample.go
package pragmastat

import (
	"fmt"
	"math"
	"sort"
	"sync"
)

// Sample wraps values with optional weights and a measurement unit.
//
// A Sample carries no error "subject": the subject (x/y/misrate) is purely a
// property of argument POSITION and is supplied by the validating function (the
// raw slice-based impls), never stored here.
type Sample struct {
	values       []float64
	weights      []float64 // nil for unweighted
	unit         *MeasurementUnit
	isWeighted   bool
	totalWeight  float64
	weightedSize float64

	sortCache *sortedCache
}

// sortedCache holds the lazily-computed sorted values/weights. It lives behind a
// pointer so the Sample struct stays copy-safe (the sync.Once is not embedded in
// Sample) and so a unit-converted view can keep its own cache.
type sortedCache struct {
	once    sync.Once
	values  []float64
	weights []float64
}

// Values returns a defensive copy of the sample values.
func (s *Sample) Values() []float64 {
	result := make([]float64, len(s.values))
	copy(result, s.values)
	return result
}

// Weights returns a defensive copy of the sample weights, or nil if unweighted.
func (s *Sample) Weights() []float64 {
	if s.weights == nil {
		return nil
	}
	result := make([]float64, len(s.weights))
	copy(result, s.weights)
	return result
}

// Unit returns the measurement unit of the sample.
func (s *Sample) Unit() *MeasurementUnit { return s.unit }

// IsWeighted returns true if the sample has weights.
func (s *Sample) IsWeighted() bool { return s.isWeighted }

// TotalWeight returns the total weight of the sample.
func (s *Sample) TotalWeight() float64 { return s.totalWeight }

// WeightedSize returns the effective sample size accounting for weights.
func (s *Sample) WeightedSize() float64 { return s.weightedSize }

// NewSample creates an unweighted sample from numeric values.
func NewSample[T Number](values []T) (*Sample, error) {
	return NewSampleWithUnit(values, nil)
}

// NewSampleWithUnit creates an unweighted sample with a specified unit.
func NewSampleWithUnit[T Number](values []T, unit *MeasurementUnit) (*Sample, error) {
	return newSample(values, nil, unit)
}

// NewWeightedSample creates a weighted sample.
func NewWeightedSample[T Number](values []T, weights []float64, unit *MeasurementUnit) (*Sample, error) {
	return newSample(values, weights, unit)
}

// newSample constructs a Sample, validating the values. Construction validity
// errors (empty / NaN / Inf) are always reported with subject "x": construction
// cannot know which argument position the sample will occupy.
func newSample[T Number](values []T, weights []float64, unit *MeasurementUnit) (*Sample, error) {
	if unit == nil {
		unit = NumberUnit
	}
	if len(values) == 0 {
		return nil, NewValidityError(SubjectX)
	}

	fValues := make([]float64, len(values))
	for i, v := range values {
		fv := float64(v)
		if math.IsNaN(fv) || math.IsInf(fv, 0) {
			return nil, NewValidityError(SubjectX)
		}
		fValues[i] = fv
	}

	s := &Sample{
		values:    fValues,
		unit:      unit,
		sortCache: &sortedCache{},
	}

	if weights != nil {
		if len(weights) != len(values) {
			return nil, fmt.Errorf("weights length (%d) must match values length (%d)", len(weights), len(values))
		}
		var totalWeight, totalWeightSq float64
		var minW = math.MaxFloat64
		for _, w := range weights {
			totalWeight += w
			totalWeightSq += w * w
			if w < minW {
				minW = w
			}
		}
		if minW < 0 {
			return nil, fmt.Errorf("all weights must be non-negative")
		}
		if totalWeight < 1e-9 {
			return nil, fmt.Errorf("total weight must be positive")
		}
		wCopy := make([]float64, len(weights))
		copy(wCopy, weights)
		s.weights = wCopy
		s.isWeighted = true
		s.totalWeight = totalWeight
		s.weightedSize = (totalWeight * totalWeight) / totalWeightSq
	} else {
		s.isWeighted = false
		s.totalWeight = 1.0
		s.weightedSize = float64(len(fValues))
	}

	return s, nil
}

// Size returns the number of values.
func (s *Sample) Size() int {
	return len(s.values)
}

// SortedValues returns a sorted copy of the values (lazily computed).
func (s *Sample) SortedValues() []float64 {
	sorted := s.cachedSortedValues()
	result := make([]float64, len(sorted))
	copy(result, sorted)
	return result
}

// cachedSortedValues returns the internal sorted values cache without copying.
// Package-internal use only — callers must not mutate the returned slice.
func (s *Sample) cachedSortedValues() []float64 {
	s.ensureSorted()
	return s.sortCache.values
}

// ensureSorted lazily computes the sorted cache exactly once, even when several
// goroutines share this *Sample. sync.Once also establishes the happens-before
// so every caller observes the written cache.
func (s *Sample) ensureSorted() {
	s.sortCache.once.Do(s.computeSorted)
}

func (s *Sample) computeSorted() {
	n := len(s.values)
	sortedValues := make([]float64, n)
	copy(sortedValues, s.values)

	if s.isWeighted && s.weights != nil {
		sortedWeights := make([]float64, n)
		copy(sortedWeights, s.weights)
		// Co-sort values and weights
		indices := make([]int, n)
		for i := range indices {
			indices[i] = i
		}
		sort.Slice(indices, func(i, j int) bool {
			return sortedValues[indices[i]] < sortedValues[indices[j]]
		})
		tmpV := make([]float64, n)
		tmpW := make([]float64, n)
		for i, idx := range indices {
			tmpV[i] = sortedValues[idx]
			tmpW[i] = sortedWeights[idx]
		}
		copy(sortedValues, tmpV)
		copy(sortedWeights, tmpW)
		s.sortCache.weights = sortedWeights
	} else {
		sort.Float64s(sortedValues)
	}
	s.sortCache.values = sortedValues
}

// ConvertTo converts the sample to a different (compatible) unit.
func (s *Sample) ConvertTo(target *MeasurementUnit) (*Sample, error) {
	if !s.unit.IsCompatible(target) {
		return nil, &UnitMismatchError{Unit1: s.unit, Unit2: target}
	}
	if s.unit == target {
		return s, nil
	}
	factor := ConversionFactor(s.unit, target)
	converted := make([]float64, len(s.values))
	for i, v := range s.values {
		converted[i] = v * factor
	}
	result := &Sample{
		values:       converted,
		unit:         target,
		isWeighted:   s.isWeighted,
		totalWeight:  s.totalWeight,
		weightedSize: s.weightedSize,
		sortCache:    &sortedCache{},
	}
	if s.isWeighted {
		result.weights = make([]float64, len(s.weights))
		copy(result.weights, s.weights)
	}
	return result, nil
}

// checkNonWeighted returns an error if the sample is weighted.
func checkNonWeighted(name string, s *Sample) error {
	if s == nil {
		return fmt.Errorf("%s cannot be nil", name)
	}
	if s.isWeighted {
		return fmt.Errorf("weighted samples are not supported for %s", name)
	}
	return nil
}

// checkCompatibleUnits returns an error if two samples have incompatible units.
func checkCompatibleUnits(a, b *Sample) error {
	if !a.unit.IsCompatible(b.unit) {
		return &UnitMismatchError{Unit1: a.unit, Unit2: b.unit}
	}
	return nil
}

// convertToFiner converts both samples to the finer unit.
func convertToFiner(a, b *Sample) (*Sample, *Sample, error) {
	if a.unit == b.unit {
		return a, b, nil
	}
	target := Finer(a.unit, b.unit)
	newA, err := a.ConvertTo(target)
	if err != nil {
		return nil, nil, err
	}
	newB, err := b.ConvertTo(target)
	if err != nil {
		return nil, nil, err
	}
	return newA, newB, nil
}

File: ts/src/sample.ts
/**
 * Sample wraps values with optional weights and a measurement unit.
 */

import { MeasurementUnit } from './measurement-unit';
import { AssumptionError } from './assumptions';
export class Sample {
  readonly values: readonly number[];
  readonly weights: readonly number[] | null;
  readonly unit: MeasurementUnit;
  readonly isWeighted: boolean;
  readonly totalWeight: number;
  readonly weightedSize: number;

  private _sortedValues: number[] | null = null;

  constructor(values: readonly number[], weights: readonly number[] | null, unit: MeasurementUnit) {
    if (values.length === 0) {
      throw AssumptionError.validity('x');
    }
    for (const v of values) {
      if (!Number.isFinite(v)) {
        throw AssumptionError.validity('x');
      }
    }

    this.values = values;
    this.unit = unit;

    if (weights !== null) {
      if (weights.length !== values.length) {
        throw new Error(
          `weights length (${weights.length}) must match values length (${values.length})`,
        );
      }
      let totalWeight = 0;
      let totalWeightSq = 0;
      for (const w of weights) {
        if (w < 0) {
          throw new Error('all weights must be non-negative');
        }
        totalWeight += w;
        totalWeightSq += w * w;
      }
      if (totalWeight < 1e-9) {
        throw new Error('total weight must be positive');
      }
      this.weights = weights;
      this.isWeighted = true;
      this.totalWeight = totalWeight;
      this.weightedSize = (totalWeight * totalWeight) / totalWeightSq;
    } else {
      this.weights = null;
      this.isWeighted = false;
      this.totalWeight = 1.0;
      this.weightedSize = values.length;
    }
  }

  /** Number of values in the sample. */
  get size(): number {
    return this.values.length;
  }

  /** Returns a sorted copy of the values (lazily computed). */
  get sortedValues(): number[] {
    if (this._sortedValues === null) {
      this._sortedValues = [...this.values].sort((a, b) => a - b);
    }
    return this._sortedValues;
  }

  /** Creates an unweighted sample from numeric values. */
  static of(values: number[]): Sample {
    return new Sample(values, null, MeasurementUnit.NUMBER);
  }

  /** Creates an unweighted sample with a specified unit. */
  static withUnit(values: number[], unit: MeasurementUnit): Sample {
    return new Sample(values, null, unit);
  }

  /** Creates a weighted sample with an optional unit. */
  static weighted(
    values: number[],
    weights: number[],
    unit: MeasurementUnit = MeasurementUnit.NUMBER,
  ): Sample {
    return new Sample(values, weights, unit);
  }

  /** Converts the sample to a different (compatible) unit. */
  convertTo(target: MeasurementUnit): Sample {
    if (!this.unit.isCompatible(target)) {
      throw new AssumptionError(`can't convert ${this.unit.fullName} to ${target.fullName}`);
    }
    if (this.unit === target) {
      return this;
    }
    const factor = MeasurementUnit.conversionFactor(this.unit, target);
    const converted = this.values.map((v) => v * factor);
    return new Sample(converted, this.weights !== null ? [...this.weights] : null, target);
  }

  /** Returns a new sample with log-transformed values and NumberUnit. */
  log(): Sample {
    const logValues = new Array<number>(this.values.length);
    for (let i = 0; i < this.values.length; i++) {
      if (this.values[i] <= 0) {
        throw AssumptionError.positivity('x');
      }
      logValues[i] = Math.log(this.values[i]);
    }
    return new Sample(
      logValues,
      this.weights !== null ? [...this.weights] : null,
      MeasurementUnit.NUMBER,
    );
  }
}

/** Checks that a sample is not weighted; throws AssumptionError if it is. */
export function checkNonWeighted(s: Sample): void {
  if (s.isWeighted) {
    throw new AssumptionError('weighted samples are not supported');
  }
}

/** Checks that two samples have compatible units; throws AssumptionError if not. */
export function checkCompatibleUnits(a: Sample, b: Sample): void {
  if (!a.unit.isCompatible(b.unit)) {
    throw new AssumptionError(`can't convert ${a.unit.fullName} to ${b.unit.fullName}`);
  }
}

/** Converts both samples to their finer unit. */
export function convertToFiner(a: Sample, b: Sample): [Sample, Sample] {
  if (a.unit === b.unit) {
    return [a, b];
  }
  const target = MeasurementUnit.finer(a.unit, b.unit);
  return [a.convertTo(target), b.convertTo(target)];
}
